added an 'openmdao graph' command

naylor-b opened the pull request:
### Summary

User can now, via the `openmdao graph` command line tool or programmatically by using the `write_graph` method on a Group, generate graphs viewable in a web browser.  Various options control whether variables are displayed, whether Groups are collapsed, and whether the graph is a dataflow graph or a tree.

### Backwards incompatibilities

None

### New Dependencies

This particular command depends on the `pydot` package, but `pydot` was not added as an OpenMDAO dependency.

swryan commented:
Instead of adding so many additional functions to Group, I wonder if this is a good candidate for encapsulation in it's own module/class...  i.e. put all those new functions into `visualization/graph_viewer.py` and make it a class so it could be invoked as   `GraphViewer(group).write_graph(options...)`

It seems like that might be a little cleaner?

swryan commented on openmdao/visualization/graph_viewer.py:
```diff
@@ -590,6 +590,9 @@ def write_graph(G, prog='dot', display=True, outfile='graph.svg'):
 
+    if ext == 'svg':
+        outfile = outfile.replace('.svg', '.html')
+
```
this might be a little confusing.. since the output file is different than what was requested..  
would it be better if the default extension is html, which uses `create-svg` internally?

swryan commented on openmdao/visualization/graph_viewer.py:
```diff
@@ -0,0 +1,796 @@
+        """
+        if pydot is None:
+            issue_warning("To view the system graph, you must install pydot. "
+                          "You can install it via 'pip install pydot'.")
```
Sorry about the piecemeal comments..  

I would suggest adding a mention that `pydot` also requires `graphviz`...  e.g.
```
   issue_warning("To view the system graph, you must install pydot. "
                          "You can install it via 'pip install pydot'.\n"
                          "Note that pydot requires graphviz, which is a non-Python application.\n"
                          "It can be installed at the system level or via a package manager like conda ")
```

```
[11:13:55](latest) swryan@Chimaera:~/dev/OpenMDAO/openmdao/test_suite/scripts$ openmdao graph beam_opt.py
/home/swryan/dev/OpenMDAO/openmdao/visualization/graph_viewer.py:103: OpenMDAOWarning:To view the system graph, you must install pydot. You can install it via 'pip install pydot'.
[11:14:08](latest) swryan@Chimaera:~/dev/OpenMDAO/openmdao/test_suite/scripts$ pip install pydot
Collecting pydot
  Downloading pydot-2.0.0-py3-none-any.whl.metadata (9.6 kB)
Requirement already satisfied: pyparsing>=3 in /home/swryan/miniconda3/envs/latest/lib/python3.12/site-packages (from pydot) (3.1.2)
Downloading pydot-2.0.0-py3-none-any.whl (22 kB)
Installing collected packages: pydot
Successfully installed pydot-2.0.0
[11:14:15](latest) swryan@Chimaera:~/dev/OpenMDAO/openmdao/test_suite/scripts$ openmdao graph beam_opt.py
Traceback (most recent call last):
  File "/home/swryan/miniconda3/envs/latest/lib/python3.12/site-packages/pydot/core.py", line 1753, in create
    stdout_data, stderr_data, process = call_graphviz(
                                        ^^^^^^^^^^^^^^
  File "/home/swryan/miniconda3/envs/latest/lib/python3.12/site-packages/pydot/core.py", line 133, in call_graphviz
    process = subprocess.Popen(
              ^^^^^^^^^^^^^^^^^
  File "/home/swryan/miniconda3/envs/latest/lib/python3.12/subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/home/swryan/miniconda3/envs/latest/lib/python3.12/subprocess.py", line 1953, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'dot'

[11:14:38](latest) swryan@Chimaera:~/dev/OpenMDAO/openmdao/test_suite/scripts$ conda install graphviz
Retrieving notices: ...working... done
Channels:
 - conda-forge
 - defaults
Platform: linux-64
Collecting package metadata (repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /home/swryan/miniconda3/envs/latest

  added / updated specs:
    - graphviz


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    atk-1.0-2.38.0             |       hd4edc92_1         539 KB  conda-forge
    expat-2.6.2                |       h59595ed_0         134 KB  conda-forge
    fribidi-1.0.10             |       h36c2ea0_0         112 KB  conda-forge
    gdk-pixbuf-2.42.10         |       h829c605_5         560 KB  conda-forge
    graphite2-1.3.13           |    h59595ed_1003          95 KB  conda-forge
    graphviz-9.0.0             |       h78e8752_1         2.2 MB  conda-forge
    gtk2-2.24.33               |       h280cfa0_4         6.2 MB  conda-forge
    gts-0.7.6                  |       h977cf35_4         311 KB  conda-forge
    libdeflate-1.20            |       hd590300_0          70 KB  conda-forge
    libgd-2.3.3                |       h119a65a_9         219 KB  conda-forge
    libglib-2.80.0             |       hf2295e7_1         2.8 MB  conda-forge
    librsvg-2.56.3             |       he3f83f7_1         5.6 MB  conda-forge
    libtiff-4.6.0              |       h1dd3fc0_3         276 KB  conda-forge
    libwebp-1.3.2              |       h658648e_1          83 KB  conda-forge
    pango-1.52.1               |       ha41ecd1_0         434 KB  conda-forge
    pixman-0.43.2              |       h59595ed_0         378 KB  conda-forge
    ------------------------------------------------------------
                                           Total:        19.9 MB

The following NEW packages will be INSTALLED:

  atk-1.0            conda-forge/linux-64::atk-1.0-2.38.0-hd4edc92_1
  cairo              conda-forge/linux-64::cairo-1.18.0-h3faef2a_0
  expat              conda-forge/linux-64::expat-2.6.2-h59595ed_0
  font-ttf-dejavu-s~ conda-forge/noarch::font-ttf-dejavu-sans-mono-2.37-hab24e00_0
  font-ttf-inconsol~ conda-forge/noarch::font-ttf-inconsolata-3.000-h77eed37_0
  font-ttf-source-c~ conda-forge/noarch::font-ttf-source-code-pro-2.038-h77eed37_0
  font-ttf-ubuntu    conda-forge/noarch::font-ttf-ubuntu-0.83-h77eed37_1
  fontconfig         conda-forge/linux-64::fontconfig-2.14.2-h14ed4e7_0
  fonts-conda-ecosy~ conda-forge/noarch::fonts-conda-ecosystem-1-0
  fonts-conda-forge  conda-forge/noarch::fonts-conda-forge-1-0
  freetype           conda-forge/linux-64::freetype-2.12.1-h267a509_2
  fribidi            conda-forge/linux-64::fribidi-1.0.10-h36c2ea0_0
  gdk-pixbuf         conda-forge/linux-64::gdk-pixbuf-2.42.10-h829c605_5
  gettext            conda-forge/linux-64::gettext-0.21.1-h27087fc_0
  giflib             conda-forge/linux-64::giflib-5.2.1-h0b41bf4_3
  graphite2          conda-forge/linux-64::graphite2-1.3.13-h59595ed_1003
  graphviz           conda-forge/linux-64::graphviz-9.0.0-h78e8752_1
  gtk2               conda-forge/linux-64::gtk2-2.24.33-h280cfa0_4
 
[... DELETED ...]

Proceed ([y]/n)?


Downloading and Extracting Packages:

Preparing transaction: done
Verifying transaction: done
Executing transaction: \
/
done
[11:15:31](latest) swryan@Chimaera:~/dev/OpenMDAO/openmdao/test_suite/scripts$ openmdao graph beam_opt.py
```

naylor-b commented:
Added text that mentions dependency on pydot and graphviz and changed warning to exception.  I agree that we are inconsistent about using `view_` in our commands, but they're split fairly equally so I'd rather not change a bunch of them on this branch.
